src/main/java/com/dimensiondelvers/dimensiondelvers/server/inventorySnapshot/InventorySnapshotSystem.java

```diff
@@ -0,0 +1,210 @@
+            for (ItemEntity itemEntity : drops) {
+                ItemStack item = itemEntity.getItem();
+
+                if (item.getComponents().has(ModDataComponentType.INVENTORY_SNAPSHOT_ID.get()) && snapshotItemIds.contains(item.getComponents().get(ModDataComponentType.INVENTORY_SNAPSHOT_ID.get()))) {
+                    if (item.has(DataComponents.CONTAINER)) {
+                        processContainerContents(item, true);
+                    }
+                    retainItems.add(item);
+                } else if (item.isStackable()) {
+                    int dropCount = calculateDropCount(item);
+
+                    if (dropCount < item.getCount()) {
+                        retainItems.add(item.split(item.getCount() - dropCount));
+                    }
+                    if (!item.isEmpty()) {
+                        dropItems.add(itemEntity);
+                    }
+                } else {
+                    if (item.has(DataComponents.CONTAINER)) {
+                        processContainerContents(item, false);
+                    }
+                    itemEntity.getItem().applyComponents(REMOVE_SNAPSHOT_ID_PATCH);
+                    dropItems.add(itemEntity);
+                }
+            }
+        }
+
+        // If we're retaining the container
+        // - Any item that should be retained keep in container
+        // - Any item that we don't want, copy and clear and drop the copy
+        // If we're not retaining the container
+        // - Any item that should be retained copy and clear and put the copy in retain
+        // - Any item that we don't want, keep in container
+        private void processContainerContents(ItemStack containerItem, boolean retainingContainer) {
+            ItemContainerContents itemContainerContents = containerItem.get(DataComponents.CONTAINER);
+            for (ItemStack item : itemContainerContents.nonEmptyItems()) {
+                if (item.getComponents().has(ModDataComponentType.INVENTORY_SNAPSHOT_ID.get()) && snapshotItemIds.contains(item.getComponents().get(ModDataComponentType.INVENTORY_SNAPSHOT_ID.get()))) {
+                    if (item.has(DataComponents.CONTAINER)) {
+                        processContainerContents(item, true);
+                    }
+
+                    if (!retainingContainer) {
+                        retainItems.add(item.copyAndClear());
+                    }
+                } else if (item.isStackable()) {
+                    int dropCount = calculateDropCount(item);
+
+                    if (retainingContainer && dropCount > 0) {
+                        dropItems.add(createItemEntity(item.split(dropCount)));
+                    } else if (!retainingContainer && dropCount < item.getCount()) {
+                        retainItems.add(item.split(item.getCount() - dropCount));
+                    }
+                } else if (retainingContainer) {
+                    item.applyComponents(REMOVE_SNAPSHOT_ID_PATCH);
```

Comment:
There is a missing check for if the item is a container and processing of its contents here.